### PR TITLE
grouping for types

### PR DESCRIPTION
**Pull request by FredTheDino**
- Format Void in unions as ?
- Add really good test
- Agressive test
- Make things PartialEq
- Simplify type
- Not today!
- Chicken out of the change and keep what can be kept
- TypeKind::Grouping
- Solve the parsing problems
- Fix some tests

### CODE REVIEW

File: sylt-parser/src/parser.rs [hunk 1]
```diff
@@ -182,6 +182,8 @@ pub enum TypeKind {
     Dict(Box<Type>, Box<Type>),
     /// A generic type
     Generic(Identifier),
+    /// `(inner_type)` - usefull to correct the grouping
```
Comment (EMattfolk) [inline sylt-parser/src/parser.rs:185]:
```suggestion
    /// `(inner_type)` - useful for correcting ambiguous types
```

File: sylt-parser/src/parser.rs [hunk 2]
```diff
@@ -539,33 +541,33 @@ pub fn parse_type<'t>(ctx: Context<'t>) -> ParseResult<'t, Type> {
             let mut ctx = ctx.skip(1);
             let mut types = Vec::new();
             // Tuples may (and probably will) contain multiple types.
+            let mut is_tuple = matches!(ctx.token(), T::Comma | T::RightParen);
```
Comment (EMattfolk) [inline sylt-parser/src/parser.rs:544]: It shouldn't be able to start with a comma right?
Perhaps skip it at the end of the loop.

File: sylt-parser/src/parser.rs [hunk 3]
```diff
@@ -1166,8 +1170,8 @@ impl Display for Type {
             TypeKind::Tuple(tys) => {
                 write!(f, "(")?;
                 for (i, ty) in tys.iter().enumerate() {
-                    if i != 0 { write!(f, ", ")?; }
-                    write!(f, "{}", ty)?;
+                    if i != 0 { write!(f, " ")?; }
+                    write!(f, "{},", ty)?;
```
Comment (EMattfolk) [inline sylt-parser/src/parser.rs:1174]: I think this causes trailing commas.

File: sylt-parser/src/parser.rs (hunk 2 above)
Comment (FredTheDino) [reply sylt-parser/src/parser.rs:544]: It now matches the tuple syntax for values. As far as I know, this is correct.

File: sylt-parser/src/parser.rs (hunk 3 above)
Comment (FredTheDino) [reply sylt-parser/src/parser.rs:1174]: Yes, but it's acceptable in tuples, I made a formatting decision that you want trailing commas in tuples.

Comment (FredTheDino) [reply sylt-parser/src/parser.rs:1174]: It simplifies the implementation for the debug prints. I don't care enough about it to add a special case. If you feel strongly make an issue.

**Pull request merged**